app/controllers/spree/orders_controller_decorator.rb

```diff
@@ -202,4 +223,23 @@ def populate_cart hash
   def wrap_json_infinity(n)
     n == Float::INFINITY ? 2147483647 : n
   end
+
+  def order_to_update
+    order = Spree::Order.complete.find_by_number(params[:id])
+    return order if order.andand.changes_allowed? && can?(:update, order)
+    current_order
+  end
```

Comment:
Shouldn't this logic go into the authorization layer? CanCan abilities for instance? What's the difference between `current_order` and `order` here?